select-project component

dtaylor113 opened the pull request:
Created a `<select-project>` component which allows one to select or create a new project.  Integrated into the serviceClass order service:

![image](https://cloud.githubusercontent.com/assets/12733153/24723208/a021c804-1a14-11e7-9cca-9f852518875e.png)
The most recently created project is selected by default.

**Project keyword search:**
![image](https://cloud.githubusercontent.com/assets/12733153/24723245/cb2e44dc-1a14-11e7-8f6a-61d8a4b4ddbd.png)

**Create Project with validation:**
![image](https://cloud.githubusercontent.com/assets/12733153/24723285/f5744f7a-1a14-11e7-9c8a-a1181f6b74c1.png)
Duplicate project name validation also implemented.

spadgett commented:
Suggest we should remove the help text from below the project name field since we're not adding it to other fields (at least not yet).

Screenshots look great!

dtaylor113 commented:
> Suggest we should remove the help text from below the project name field since we're not adding it to other fields (at least not yet).

That is the standard which the Project Summary uses and `<create-project>` dialog uses.

dtaylor113 commented:
Also pls note that I haven't used the `<select-project>` component nor the serviceClass order service against real data.  I plan on setting up my system tomorrow to do this.

dtaylor113 commented:
I'm also seeing a highlight issue when selecting a project:

![image](https://cloud.githubusercontent.com/assets/12733153/24724602/b2d65c6c-1a19-11e7-9857-e2b473ee4e0b.png)

I believe we are getting this because we are using a `<ui-select>` within a `<form>` I spent some time trying to override it in less/css but couldn't find a solution.  @rhamilto I will create an issue if I can't find a solution.  -thanks

spadgett commented on src/components/order-service/order-service.controller.ts:
```diff
@@ -154,4 +180,8 @@ export class OrderServiceController implements angular.IController {
   }
+
+  private isNewProject(): boolean {
+    return this.ctrl.selectedProject && this.ctrl.selectedProject.metadata ? this.ctrl.selectedProject.metadata.uid === 'create_project' : false;
```
I'd rather leave `uid` undefined instead of setting it to something else. Then a project without a `uid` is new.

I know this will never actually happen, but technically `create_project` is a valid UID for a real project :)

spadgett replied:
Maybe

```js
return this.ctrl.selectedProject && !_.has(this.ctrl.selectedProject, 'metadata.uid');
```

spadgett replied:
An alternative is to have a property like `newProjectOption: true`

dtaylor113 replied:
Fixed.  Using an undefined `uid` to determine if creating a new project.

spadgett commented on src/components/order-service/order-service.html:
```diff
@@ -75,7 +75,7 @@
         class="btn btn-primary wizard-pf-next"
         ng-if="$ctrl.currentStep.id === 'configure'"
-        ng-disabled="!$ctrl.selectedProject || $ctrl.currentStep.id === 'results'"
+        ng-disabled="!$ctrl.selectedProject || $ctrl.currentStep.id === 'results' || $ctrl.invalidNewProject()"
```
Better to have a nested `ng-form` for new project and make that form invalid. Then we can check form validity similar to 

https://github.com/openshift/origin-web-catalog/pull/78/files#diff-a816c7d825f3b359053f0814dcc271f2R32

dtaylor113 replied:
Hi @spadgett, the new project form is in the select project component, not in order service html.  Not sure what you are suggesting here.

spadgett replied:
You can nest forms using `ng-form`.

https://docs.angularjs.org/api/ng/directive/ngForm

I'm suggesting the order service configuration page has a form. The select project component uses ng-form. Then if the select project form is invalid, the outer order service form will be marked invalid. Then we disable the button whenever the form is invalid.

We will need form validation anyway when we add parameters.

spadgett replied:
Here's an example of how you can use ng-form:

https://github.com/openshift/origin-web-console/blob/95b0902ecb3627301fedf2a0a6855cbd5e8c39a2/app/views/directives/select-containers.html

Then we embedded the `select-container` directive in a larger form, which was marked invalid if a container wasn't selected.

dtaylor113 replied:
Fixed.

spadgett commented on src/components/select-project/select-project.controller.ts:
```diff
@@ -0,0 +1,93 @@
+  public $onInit() {
+    this.ctrl.nameTaken = false;
+    // TODO disableInputs?
+    this.ctrl.disableInputs = false;
```
I don't think this is needed, or at least it should be on the entire form, not just the select project control.

dtaylor113 replied:
Fixed. Removed

spadgett commented on src/components/select-project/select-project.controller.ts:
```diff
@@ -0,0 +1,93 @@
+
+    this.ProjectsService.canCreate().then(() => {
+      this.ctrl.canCreate = true;
+      this.AuthService.withUser().then(this.onVerifyUser);
```
`withUser` shouldn't be needed. You should be logged in at this point. You would need to call it before `canCreate()` anyway.

dtaylor113 replied:
Ok, this is very similar to how the project summary is doing it.  Should I also change it here:  https://github.com/openshift/origin-web-catalog/blob/master/src/components/projects-summary/projects-summary.controller.ts#L36
-thanks

spadgett replied:
I think so. You can't properly check `canCreate` without being logged in first.

dtaylor113 replied:
Fixed in both places.

spadgett commented on src/components/select-project/select-project.controller.ts:
```diff
@@ -0,0 +1,93 @@
+        }
+        this.Logger.warn(msg);
+      }
+      this.AuthService.withUser().then(this.onVerifyUser);
```
Code shared by both success and failure can use the `finally` method.

dtaylor113 replied:
Fixed

spadgett commented on src/components/select-project/select-project.controller.ts:
```diff
@@ -0,0 +1,93 @@
+      this.ctrl.projects = _.sortBy(response.by('metadata.name'), this.$filter('displayName'));
+
+      // get most recently created
+      this.ctrl.selectedProject = _.first(this.$filter('orderObjectsByDate')(this.ctrl.projects, true));
```
You can loop through the projects once -- O(n) -- to find most recent rather than sorting -- O(n log(n)). It's possible for admins to see hundreds or thousands of projects, so it can make a difference.

Here is an example:

https://github.com/openshift/origin-web-console/blob/8f4cda159d1382b7e5e2d4c3223b4ac248be6e1f/app/scripts/filters/date.js#L99-L110

dtaylor113 replied:
Waiting on origin-web-common PR to gain access to those filters.

dtaylor113 replied:
Fixed.

rhamilto commented:
>  @rhamilto I will create an issue if I can't find a solution.

That may be a global console bug.  Please open an issue.

spadgett commented:
@dtaylor113 Do you see the ui-select bug in the console as well? I noticed noticed the caret doesn't look positioned properly in your screenshots.

spadgett commented on src/components/select-project/selectProject.html:
```diff
@@ -0,0 +1,93 @@
+    <div class="form-group">
+      <label for="name" class="col-sm-4 control-label required">Project Name</label>
+      <span class="col-sm-8" ng-class="{'has-error': ($ctrl.createProjectForm.name.$error.pattern && $ctrl.createProjectForm.name.$touched) || nameTaken}">
+        <input class="form-control input-lg"
```
We shouldn't be using `input-lg`

dtaylor113 replied:
For all the input fields or just name?

spadgett replied:
For all inputs. Let's use consistent input sizes on this form.

jeff-phillips-18 commented on src/components/select-project/select-project.controller.ts:
```diff
@@ -0,0 +1,93 @@
+
+  public $onChanges(onChangesObj: angular.IOnChangesObject) {
+    if (onChangesObj.nameTaken && !onChangesObj.nameTaken.isFirstChange()) {
+      this.ctrl.nameTaken = onChangesObj.nameTaken.currentValue;
```
this is not necessary, it will already have been assigned.

dtaylor113 replied:
Fixed

jeff-phillips-18 commented:
I think it might be better to have the existing projects list, then a + button to the left which simply pops up the create project panel (like in the projects summary panel). That would remove the redundancy of creating  a project panel here.

dtaylor113 commented on src/components/order-service/order-service.controller.ts:
```diff
@@ -111,40 +149,28 @@ export class OrderServiceController implements angular.IController {
+    spec = planName ? _.assign(spec, {planName: planName}) : spec;
+
+    let metadata: any = {generateName: serviceClassName + '-'};
+    metadata = projectName ? _.assign(metadata, {namespace: projectName}) : metadata;
```
Put back original code, updated mockData to always have at least one plan.

dtaylor113 commented:
> That may be a global console bug. Please open an issue.

#87

dtaylor113 commented:
> @dtaylor113 Do you see the ui-select bug in the console as well?

@spadgett I'm having trouble viewing the 'create from url' (which has a ui-select) in console.

dtaylor113 commented:
> I think it might be better to have the existing projects list, then a + button to the right which simply pops up the create project panel (like in the projects summary panel). That would remove the redundancy of creating a project panel here.

@jeff-phillips-18 Sorry, I don't follow.  Are you just agreeing with the design here?

spadgett commented:
> That is the standard which the Project Summary uses and <create-project> dialog uses. I can remove it, but it won't be consistent with other usages.

It's more important to be consistent with fields on the same form. They don't have the `input-lg` class, placeholder text, or field help currently. The new project form will show up in the dialog from #78 and also alongside service instance parameters when we add them.

If we decide we want help and placeholders, I think we need to go through and add them everywhere.

jeff-phillips-18 commented:
I'm kind of disagreeing with the design. I'm suggesting take the 'Create new project' out of the dropdown choices. Add a + button to the right of the combo box which when clicks the + button we simply show the create project bubble to allow the user to create a project (just like the Create New Project button in the sidebar). Once created, auto-select the project in the list.

spadgett commented:
I personally like that it all feels like one action to the user. Having multiple forms and dialogs on top of dialogs doesn't seem as clean to me, even if implementation is harder this way.

spadgett commented:
@dtaylor113 Thanks for your patience with all my comments. I really like the functionality.

jeff-phillips-18 commented:
Maybe I'd like it better if the form didn't show up only after selecting "Create Project". I feel that without the form the page is very empty.

spadgett commented:
> I feel that without the form the page is very empty.

Yeah, I agree, and I mentioned this to @serenamarie125 as well. Eventually we'll have parameters here, too, but only for service classes that use them.

dtaylor113 commented:
> It's more important to be consistent with fields on the same form. They don't have the input-lg class, placeholder text, or field help currently.

@spadgett Agree to remove ` input-lg ` so all form fields look the same.  I'd rather not remove the placeholder or help text as I believe some thought went into making the form easy to use.  I vote for adding placeholders and help text (if appropriate) to the other forms fields.

spadgett commented on src/components/order-service/order-service.controller.ts:
```diff
@@ -94,7 +96,44 @@ export class OrderServiceController implements angular.IController {
+        }, (result: any) => {
+          var data = result.data || {};
+          if (data.reason === 'AlreadyExists') {
+            this.ctrl.nameTaken = true;
```
Hey @dtaylor113 I was thinking since we already have the list of projects, can we check name taken before trying to create the project? We had a directive that does something like this in origin-web-console.

https://github.com/openshift/origin-web-console/blob/e7a0c0a8f703d5429f70b78223abb31856a66670/app/scripts/directives/oscUnique.js

Doesn't have to be in this PR

spadgett replied:
You'll still need to check the return status as well since you probably can't see all projects.

dtaylor113 replied:
Sounds good.  #88 so we don't loose track of it.

spadgett commented:
@dtaylor113 I'm seeing the same visual problems as you. I think you might be missing the ui-select CSS.

dtaylor113 commented:
I think I've addressed all review issues except the css/styling one, which I'm going to look into some more.   There might be some lingering design considerations, which we should file issues for.  -thanks

spadgett commented on src/components/select-project/selectProject.html:
```diff
@@ -0,0 +1,93 @@
+<form class="form-horizontal config-form">
```
Needs to be `ng-form` since you can't nest forms in HTML.

```html
<ng-form class=form-horizontal config-form">
```

dtaylor113 replied:
Fixed

spadgett commented on src/components/select-project/selectProject.html:
```diff
@@ -0,0 +1,93 @@
+    </div>
+</form>
+
+<ng-form name="$ctrl.forms.createProjectForm" class="form-horizontal"
```
Just have one `ng-form` for the whole component instead of two.

dtaylor113 replied:
The top form is selectProject whereas the bottom form is createProject.  Did change both to `ng-form`

spadgett commented on src/components/select-project/selectProject.html:
```diff
@@ -0,0 +1,93 @@
+
+<ng-form name="$ctrl.forms.createProjectForm" class="form-horizontal"
+      ng-if="$ctrl.isNewProject()">
+  <fieldset>
```
`fieldset` shouldn't be needed since we removed `disableInputs`. That's why it was there.

dtaylor113 replied:
Fixed

spadgett commented on src/components/order-service/order-service.controller.ts:
```diff
@@ -94,7 +96,44 @@ export class OrderServiceController implements angular.IController {
+    if (this.isNewProject()) {
+      // the selectedProject is actually a newProject object
+      let newProjName = this.ctrl.selectedProject.metadata.name;
+      let newProjDisplayName = this.ctrl.selectedProject.metadata.annotations['new-display-name'];
```
Can we just use the normal annotation for display name `openshift.io/display-name`? It looks like you use the normal description annotation already.

dtaylor113 replied:
In the case where a user is creating a new project, the 'selected project' is the createNewProjObj used in the dropdown which has it's `openshift.io/display-name` set to 'Create Project'.  'new-display-name' contains user input for the new proj's displayName.

spadgett commented on src/components/order-service/order-service.controller.ts:
```diff
@@ -111,40 +150,25 @@ export class OrderServiceController implements angular.IController {
+
     let serviceInstance = {
       kind: 'Instance',
-      apiVersion: 'servicecatalog.k8s.io/v1alpha1',
```
Why did you take out the api version? That's required

dtaylor113 replied:
whoops! sorry, restored

spadgett replied:
Thanks :)

dtaylor113 commented:
> @dtaylor113 I'm seeing the same visual problems as you. I think you might be missing the ui-select CSS.

When I inspect the ui-select dropdown I do see `ui-select` classes:

```
.ui-select-bootstrap .ui-select-match-text span {
    display: inline-block;
    width: 100%;
    overflow: hidden;
}

catalogs-app.css:1611
* {
    box-sizing: border-box;
}
```

spadgett commented:
> When I inspect the ui-select dropdown I do see ui-select classes

I think it will look OK in origin-web-console with these styles:

https://github.com/openshift/origin-web-console/blob/master/app/styles/_select.less

spadgett commented:
@dtaylor113 As a follow on, can you pull _select.less into common so the ui-select looks correct outside origin-web-console?

spadgett commented:
Yup, thanks :)

spadgett commented on src/styles/ui-select.less:
```diff
@@ -0,0 +1,75 @@
+/* Fix some styles with the angular-ui-select bootstrap theme that isn't right when combined with all of our styles */
+.ui-select-bootstrap {
```
Let's actually put this in origin-web-common so we don't have copies of it.

dtaylor113 replied:
Ok, I removed it for now and will follow up with a PR after moving it into common.

spadgett replied:
Thanks!
